services/sidecar/rollup/derivation/batch_v0_encoder.go

```diff
@@ -0,0 +1,169 @@
+package derivation
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/specularL2/specular/services/sidecar/rollup/rpc/bridge"
+	spTypes "github.com/specularL2/specular/services/sidecar/rollup/types"
+	"github.com/specularL2/specular/services/sidecar/utils/fmt"
+	"github.com/specularL2/specular/services/sidecar/utils/log"
+)
+
+const v0 = 0x0
+
+var errBatchFull = errors.New("batch full")
+var errBatchTooSmall = errors.New("batch too small")
+
+type Config interface {
+	GetTargetBatchSize() uint64
+	GetSeqWindowSize() uint64
+	GetSubSafetyMargin() uint64
+}
+
+type BatchV0Encoder struct {
+	cfg      Config
+	timeout  uint64
+	currBuf  *bytes.Buffer
+	subBatch *subBatch
+}
+
+func NewBatchV0Encoder(cfg Config) *BatchV0Encoder {
+	return &BatchV0Encoder{cfg, 0, &bytes.Buffer{}, newSubBatch()}
+}
+
+func (e *BatchV0Encoder) GetBatch(l1Head spTypes.BlockID) ([]byte, error) {
+	if e.timeout != 0 && l1Head.GetNumber() >= e.timeout {
+		data := e.currBuf.Bytes()
+		e.currBuf.Reset()
+		return data, nil
+	}
+	if e.currBuf.Len() < int(e.cfg.GetTargetBatchSize()) {
```

Comment:
if we only check the buffer length, is it possible that we submit batches that only contain few large transactions?